lib/virtual.js

```diff
@@ -12,28 +33,73 @@ function VirtualModulesPlugin(compiler) {
 	this.compiler = compiler;
 
 	if (!compiler.inputFileSystem._writeVirtualFile) {
-		var originalPurge = compiler.inputFileSystem.purge;
+		const originalPurge = compiler.inputFileSystem.purge;
 
 		compiler.inputFileSystem.purge = function() {
 			if (originalPurge) {
-				originalPurge.call(this, arguments);
+				originalPurge.apply(this, arguments);
 			}
 			if (this._virtualFiles) {
-				Object.keys(this._virtualFiles).forEach(
-					function(file) {
-						var data = this._virtualFiles[file];
-						setData(this._statStorage, file, [null, data.stats]);
-						setData(this._readFileStorage, file, [null, data.contents]);
-					}.bind(this)
-				);
+				Object.keys(this._virtualFiles).forEach(function(file) {
+					const data = this._virtualFiles[file];
+					this._writeVirtualFile(file, data.stats, data.contents);
+				}.bind(this));
 			}
 		};
 
 		compiler.inputFileSystem._writeVirtualFile = function(file, stats, contents) {
+			const statStorage = getStatStorage(this);
+			const fileStorage = getFileStorage(this);
+			const readDirStorage = getReadDirBackend(this);
 			this._virtualFiles = this._virtualFiles || {};
 			this._virtualFiles[file] = { stats: stats, contents: contents };
-			setData(this._statStorage, file, [null, stats]);
-			setData(this._readFileStorage, file, [null, contents]);
+			setData(statStorage, file, createWebpackData(stats));
+			setData(fileStorage, file, createWebpackData(contents));
+			const segments = file.split(/[\\/]/);
+			let count = segments.length - 1;
+			const minCount = segments[0] ? 1 : 0;
+			while (count > minCount) {
```

Comment:
what is this `while` looping doing? and why is it newly required but seems to also apply to webpack 4? a code comment here would be helpful

Reply:
This is from https://github.com/sysgears/webpack-virtual-modules/commit/a73d53f8775ab3e5dd2bc57cc5e18d63408aefc8, it creates directories that the file is in.